src/cursor/common.rs

```diff
@@ -214,3 +249,43 @@ pub(crate) struct CursorInformation {
     pub(crate) batch_size: Option<u32>,
     pub(crate) max_time: Option<Duration>,
 }
+
+#[derive(Clone, Debug)]
+pub(crate) struct PinnedConnection {
```

Comment:
What do you think about modeling this as an enum instead? It avoids certain cases like `connection: None` + `valid: false`, which is nice.

e.g.
```rust
enum PinnedConnection {
    ValidPinned(Arc<Mutex<Connection>>),
    InvalidPinned(Arc<Mutex<Connection>>),
    Unpinned
}
```